### PR TITLE
Fix no-unused-variable "react" option to support self-closing JSX elements

### PR DESCRIPTION
**Pull request by adidahiya**
Also reorder some methods in NoUnusedVariableWalker.

### CODE REVIEW

File: src/rules/noUnusedVariableRule.ts [hunk 1]
```diff
@@ -121,10 +138,47 @@ class NoUnusedVariablesWalker extends Lint.RuleWalker {
...
+
+    public visitJsxElement(node: ts.JsxElement) {
+        this.hasSeenJsxElement = true;
+        super.visitJsxElement(node);
+    }
+
+    public visitJsxSelfClosingElement(node: ts.JsxSelfClosingElement) {
+        this.hasSeenJsxElement = true;
+        super.visitJsxSelfClosingElement(node);
+    }
```
Comment (adidahiya) [inline src/rules/noUnusedVariableRule.ts]: this method was added. the others were reordered.

File: test/rules/noUnusedVariableRuleTests.ts [hunk 2]
```diff
@@ -136,6 +136,10 @@ describe("<no-unused-variable>", () => {
 
             describeSuite("react");
             describeSuite("react/addons");
+
+            it("should not fail if JSX self-closing element is used", () => {
```
Comment (JKillian) [inline test/rules/noUnusedVariableRuleTests.ts]: Can we do this like the other tests for consistency?

File: test/files/rules/nounusedvariable-react4-react.test.tsx [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import * as React from "react";
+
+console.log(<input type="text" />);
```
Comment (JKillian) [inline test/files/rules/nounusedvariable-react4-react.test.tsx]: no newline at end of file

File: src/rules/noUnusedVariableRule.ts (hunk 1 above)
Comment (myitcv) [inline src/rules/noUnusedVariableRule.ts]: Thanks - this was indeed the bug. You beat me to a PR :+1:

**Comment by JKillian** [comment]
[Changes can be made in a follow-up commit, want to merge this to prepare for a release]

**Pull request merged**